README.md
# OpenCat

OpenCat is the open-source Arduino and Raspberry Pi-based quadruped robotic pet framework developed by Petoi, the maker of futuristic programmable robotic pets.

![](https://github.com/PetoiCamp/NonCodeFiles/blob/master/gif/walk.gif?raw=true)

![](https://github.com/PetoiCamp/NonCodeFiles/blob/master/gif/run.gif?raw=true)

Inspired by Boston Dynamics' Big Dog, Dr. Rongzhong Li started the project in his dorm in 2016. After one year of R&D, he founded Petoi LLC and devoted all his resources to the startup. The goal is to foster collaboration in quadruped(four-legged) robotic research, education, and engineering development of agile and affordable quadruped robot pets, bring STEM concepts to the mass and inspire newcomers (including many kids and adults) to join the robotic AI revolution to create more applications.

![](https://github.com/PetoiCamp/NonCodeFiles/blob/master/gif/slope.gif?raw=true)

The project is still a complex system only for skilled makers. Yet we want to share our design and work with the community by mass production and bring down the hardware and software costs. OpenCat has been deployed on Petoi's bionic palm-sized, realistic lifelike [cute robot cat Nybble](https://www.petoi.com/collections/robots/products/petoi-nybble-robot-cat?utm_source=github&utm_medium=code&utm_campaign=nybble) and [high-performance robot dog Bittle](https://www.petoi.com/collections/robots/products/petoi-bittle-robot-dog?utm_source=github&utm_medium=code&utm_campaign=bittle). We now have established a production line and can ship these [affordable robotic kits and accessories](https://www.petoi.com/store?utm_source=github&utm_medium=code&utm_campaign=store) worldwide.

This project provides a base open-source platform to create amazing programmable gaits, locomotion, and deployment of inverse kinematics quadruped robots and bring simulations to the real world via C/C++/Python programming languages.  Our users have deployed [NVIDIA Issac simulations and reinforcement learning on our robots](https://www.youtube.com/playlist?list=PLHMFXft_rV6MWNGyofDzRhpatxZuUZMdg). Our users have also successfully deployed OpenCat on their DIY 3D-print robot pets.

![](https://github.com/PetoiCamp/NonCodeFiles/blob/master/gif/stand.gif?raw=true)

![](https://github.com/PetoiCamp/NonCodeFiles/blob/master/gif/NybbleBalance.gif?raw=true)

We've successfully crowdfunded these two mini robot kits and shipped thousands of units worldwide.  With our customized Arduino board and servos coordinating all instinctive and sophisticated movements(walking, running, jumping, backflipping), one can clip on various sensors to bring in perception and inject artificial intelligence capabilities by mounting a Raspberry Pi or other AI chips(such as Nvidia Jetson Nano) through wired/wireless connections.  Please see [Petoi FAQs](https://www.petoi.com/pages/faq?utm_source=github&utm_medium=code&utm_campaign=faq) for more info.

Also, Check out [all of the OpenCat and Petoi robot user showcases](https://www.petoi.camp/forum/showcase).


![](https://github.com/PetoiCamp/NonCodeFiles/blob/master/gif/ball.gif?raw=true)

## Setup  Process:

OpenCat software works on both Nybble and Bittle, controlled by NyBoard based on ATmega328P. More detailed documentation can be found at the [Petoi Doc Center](https://docs.petoi.com).

To configure the board:

1. Download the repo and unfold. Remove the **-main** (or any branch name) suffix of the folder. 


2. Open the file OpenCat.ino, select your robot and board version.
```cpp
#define BITTLE    //Petoi 9 DOF robot dog: 1x on head + 8x on leg
//#define NYBBLE  //Petoi 11 DOF robot cat: 2x on head + 1x on tail + 8x on leg

//#define NyBoard_V0_1
//#define NyBoard_V0_2
#define NyBoard_V1_0
//#define NyBoard_V1_1
```

3. Comment out ```#define MAIN_SKETCH``` so that it will turn the code to the board configuration mode. Upload and follow the serial prompts to proceed. 
```cpp
// #define MAIN_SKETCH
```

4. If you activate ```#define AUTO_INIT```, the program will automatically set up without prompts. It will not reset joint offsets but calibrate the IMU. It's just a convenient option for our production line. 

5. Plug the USB uploader to the NyBoard and install the driver if no USB port is found under Arduino -> Tools -> Port.

6. Press the upload button (->) at the top-left corner in Arduino IDE.

7. Open the serial monitor of Arduino IDE. You can find the button either under Tools, or at the top-right corner of the IDE.

Set the serial monitor as **no line ending** and **115200** baud rate. 
The serial prompts:
```
Reset joint offsets? (Y/n)
Y
```

Input ‘Y’ and hit enter, if you want to reset all the joint offsets to 0. 

The program will do the reset, then update the constants and instinctive skills in the static memory. 

8. IMU (Inertial Measurement Unit) calibration.

The serial prompts:
```
Calibrate the IMU? (Y/n): 
Y 
```
Input ‘Y’ and hit enter, if you have never calibrated the IMU or want to redo calibration.

Put the robot flat on the table and don't touch it. The robot will long beep six times to give you enough time. Then it will read hundreds of sensor data and save the offsets. It will beep when the calibration finishes.

When the serial monitor prints "Ready!", you can close the serial monitor to do the next step.

9. Uncomment ```#define MAIN_SKETCH``` to make it active. This time the code becomes the normal program for the major functionalities. Upload the code.
```cpp
#define MAIN_SKETCH
```
When the serial monitor prints "Ready!", the robot is ready to take your next instructions.

10. If you have never calibrated the joints’ offsets or reset the offsets in Step2, you need to calibrate them. If you boot up the robot with one side up, it will enter the calibration state automatically for you to install the legs. Otherwise, it will enter the normal rest state

11. You can use the serial monitor to calibrate it directly. Or you may plug in the Bluetooth dongle, and use the Petoi app (on Android/iOS) for a more user-friendly interface. The mobile app is available on:

* IOS: [App Store](https://apps.apple.com/us/app/petoi/id1581548095)
* Android: [Google Play](https://play.google.com/store/apps/details?id=com.petoi.petoiapp)

You can refer to the calibration section in the user manual (https://bittle.petoi.com/6-calibration) and Guide for the Petoi App(https://docs.petoi.com/app-guide).

12. you can use the infrared remote or other applications (such as the Petoi App, python, serial monitor ... etc.) to play with the robot (https://bittle.petoi.com/7-play-with-bittle).

For updates:
* star this repository to receive timely notifications on changes.
* visit www.petoi.com and subscribe to our official newsletters for project announcements. We also host a forum at [petoi.camp](https://www.petoi.com/forum).
* follow us on [Twitter](https://twitter.com/petoicamp), [Instagram](https://www.instagram.com/petoicamp/), and [YouTube channel](https://www.youtube.com/c/rongzhongli) for fun videos and community activities.

![](https://github.com/PetoiCamp/NonCodeFiles/blob/master/gif/backflip.gif?raw=true)

[Advanced tutorials made by users](https://www.youtube.com/playlist?list=PLHMFXft_rV6MWNGyofDzRhpatxZuUZMdg)


ChangeLog.md
# Change Log

## Mar6, 2023
* Recover the skillName stored in **newCmd** array because the newCmd's memory is altered when formatting new skill data. 
* Fix the segfault caused by different string terminators: '\0', '\n', or '\~'.
* Fix the casting between int, char, and int8_t in the template function arrayNCPY().
* When turning a joint instruction into a new posture, allow large joint angles to be stored into an int8_t array. 
* Fix the positive feedback when gyro data is combined into targetAngles.
* SkillComposer: Disable the Gyro and Random buttons.
* ardSerial: Simplify the slicing algorithm when sending long commands.
* Differentiate the serial timeout for long commands over Bluetooth.
* Simplify some skills to save memory. 
* Safer auto-connection algorithm when trying to connect to a camera that's not connected to any alternative addresses. 
* Mobile app: Find the optimal delay between the mobile app's package slices to avoid overflow or timeout on the mainboard. 
* Add a token 'S' to toggle on/off the boot-up melody. 
* Add a macro to inverse the servo's rotation directions.
* Allow the robot to detect the falling-over event and recover even in turbo mode. 
* Desktop App: Add traditional Chinese and German translations. 


## Mar 1, 2023
* Combine the **newCmd** array used for receiving different kinds(ASCII/Binary, long/short) of serial commands. 
* Simplify the token cases into upper/lower letters for Binary/ASCII encodings.
* Shift the skill angle data to the end of the **newCmd** array. It allows new incoming messages to be stored at the beginning of the **newCmd** array while keeping the skill angles unchanged so that the previous skill can resume after finishing the current non-skill command. 
* It now supports much longer commands and skills, such as a behavior with **23** frames (23 * 20 + 7 = 467 bytes), a gait with (467-4)/8 = **57** frames or a long melody with 467/2 = **233** notes. On BiBoard, it supports a behavior with **125** frames (125 * 20 + 7 = 2507 bytes), a gait with (2507-4)/8 = **312** frames or a long melody with 2507/2 = **1253** notes.
* Overflow protection when the incoming message is longer than the size of **newCmd** to avoid a crash. 
* Redesign the voice commands table and their encodings. 
* Add an alias skill "up" for "balance" to save memory and convenience. 
* Move the keys of the customized voice commands to PROGMEM.
* Add new skills:

| New Skill Name | Serial Code |
|-----------|:------|
|hug      	|khg    |
|handstand  |khdsd  |
|hands up   |khu 	  |
|handshake	|khsk  	|
|sniff		  |ksnf	  |
|dig	    	|kdg	  |
|scratch    |kscrh  |
|be table 	|ktbl	  |
|boxing	  	|kbx	  |
|kick		    |kkc	  |
|come here	|kcmh	  |
|high five	|kfiv	  |
|jump		    |kjmp	  |

* With so many new skills saved to the I2C EEPROM, fix the overflow (wrapped overwriting) when a long 'K' skill data is written to the nearly full memory.
* 'T' token is used to extract the last transferred skill data. Fixed the segfault when a newly configured board has never received a 'K' command but is required to read from an unknown memory. 

## Feb 25, 2023
* Unbind the free joints' motion during walking and keep balancing reaction after the i,I,m,M commands. The free joints are unrelated to walking and balancing. By default, they move by trigonometric functions synced with the legs' pace. If a command requires specific joint movement, it will be unbound and taken over by the higher-level controller. Using the token 'i' or 'I' without more argument will re-bind them to the legs' motion.

When sending joint instructions, the rules are:


src/skill.h
#define WIRE_BUFFER 30  //Arduino wire allows a 32-byte buffer, with 2 bytes for address.
#define WIRE_LIMIT 16   //That leaves 30 bytes for data. Use 16 to balance each writes
#define PAGE_LIMIT 32   //AT24C32D 32-byte Page Write Mode. Partial Page Writes Allowed
#define EEPROM_SIZE (65535 / 8)
bool EEPROMOverflow = false;

class Skill {
public:
  byte jointIndex;
  char* skillName;  //use char array instead of String to save memory
  int8_t offsetLR;
  int8_t period;  //the period of a skill. 1 for posture, >1 for gait, <-1 for behavior
  float transformSpeed;
  byte skillHeader;
  byte frameSize;
  int expectedRollPitch[2];  //expected body orientation (roll, pitch)
  byte angleDataRatio;       //divide large angles by 1 or 2. if the max angle of skill is >128, all the angles will be divided by 2
  byte loopCycle[3];         //the looping section of a behavior (starting row, ending row, repeating cycles)
  byte firstMotionJoint;
  int8_t* dutyAngles;  //the data array for skill angles and parameters

  Skill() {
    //      period = 0;
    //      transformSpeed = 2;
    //      expectedRollPitch[0] = yprTilt[2] = 0;
    //      expectedRollPitch[1] = yprTilt[1] = 0;
    //      dutyAngles = NULL;
    //      jointIndex = 0;
  }

  void copyDataFromBufferToI2cEeprom(unsigned int eeAddress, int8_t* newCmd) {  // eeAddress should not be changed out of this function
    period = (int8_t)newCmd[0];                                                 //automatically cast to char*
    int len = dataLen(period) + 1;
    if (eeAddress + len >= EEPROM_SIZE) {
      PTLF("OVF! Won't save!");
      EEPROMOverflow = true;
#ifdef BUZZER
      beep(10, 100, 100, 5);
#endif
      return;
    }
    int writtenToEE = 0;
    while (len > 0) {
      Wire.beginTransmission(DEVICE_ADDRESS);
      Wire.write((int)((eeAddress) >> 8));  // MSB
      Wire.write((int)((eeAddress)&0xFF));  // LSB
      byte writtenToWire = 0;
      do {

        Wire.write((byte)newCmd[writtenToEE++]);
        writtenToWire++;
        eeAddress++;
      } while ((--len > 0) && (eeAddress % PAGE_LIMIT) && (writtenToWire < WIRE_LIMIT));  //be careful with the chained conditions
      //self-increment may not work as expected
      Wire.endTransmission();
      delay(6);  // needs 5ms for page write
      //PTL("\nwrote " + String(writtenToWire) + " bytes.");
    }
    //PTLF("finish copying to I2C EEPROM");
  }
#ifndef MAIN_SKETCH
  void copyDataFromPgmToI2cEeprom(unsigned int& eeAddress, unsigned int pgmAddress) {  // eeAddress should be updated after this function
    period = (int8_t)pgm_read_byte(pgmAddress);                                        //automatically cast to char*
    int len = dataLen(period) + 1;
    if (eeAddress + len >= EEPROM_SIZE) {
      PTLF("OVF! Delete some skills!");
      EEPROMOverflow = true;
#ifdef BUZZER
      beep(10, 100, 100, 5);
#endif
      return;
    }
    int writtenToEE = 0;
    while (len > 0) {
      Wire.beginTransmission(DEVICE_ADDRESS);
      Wire.write((int)((eeAddress) >> 8));  // MSB
      Wire.write((int)((eeAddress)&0xFF));  // LSB
      byte writtenToWire = 0;
      do {

        Wire.write((byte)pgm_read_byte(pgmAddress + writtenToEE++));
        writtenToWire++;
        eeAddress++;
      } while ((--len > 0) && (eeAddress % PAGE_LIMIT) && (writtenToWire < WIRE_LIMIT));  //be careful with the chained conditions
      //self-increment may not work as expected
      Wire.endTransmission();
      delay(6);  // needs 5ms for page write
      //PTL("\nwrote " + String(writtenToWire) + " bytes.");
    }
    //PTLF("finish copying to I2C EEPROM");
  }

  void saveSkillInfoFromProgmemToOnboardEeprom() {
    int skillAddressShift = 0;
    unsigned int i2cEepromAddress = INITIAL_SKILL_DATA_ADDRESS;  //won't hurt if unused
    unsigned int theLastEEaddress;
    PTLF("Save skills");
    EEPROM.update(NUM_SKILLS, sizeof(skillNameWithType) / 2);
    for (byte s = 0; s < eeprom(NUM_SKILLS); s++) {  //save skill info to on-board EEPROM
      theLastEEaddress = i2cEepromAddress;
      byte len = strlen(skillNameWithType[s]);               //len includes type. the saved value doesn't
      EEPROM.update(SKILLS + skillAddressShift++, len - 1);  //the last char in name denotes skill type, I(nstinct) on external eeprom, N(ewbility) on progmem
      // PT(skillNameWithType[s][len - 1]== 'I' ? "I nstinct\t" : "N ewbility\t");
      for (byte l = 0; l < len; l++) {
        PT(skillNameWithType[s][l]);
        EEPROM.update(SKILLS + skillAddressShift++, skillNameWithType[s][l]);
      }
      PTL();
      int8_t period = pgm_read_byte(progmemPointer[s]);
      EEPROM.update(SKILLS + skillAddressShift++, period);

      //PTL("Current EEPROM address is " + String(SKILLS + skillAddressShift));
#ifdef I2C_EEPROM
      if (!EEPROMOverflow)
        if (skillNameWithType[s][len - 1] == 'I') {  //  if there's instinct and there's i2c eeprom, and user decide to update.
          // save the data array to i2c eeprom. its address will be saved to onboard eeprom
          EEPROMWriteInt(SKILLS + skillAddressShift, i2cEepromAddress);
          copyDataFromPgmToI2cEeprom(i2cEepromAddress, (unsigned int)progmemPointer[s]);
          //            PT(int(SKILLS + skillAddressShift + 2)); PT(" I2C EE "); PTL(int(i2cEepromAddress));
        }
#endif
      skillAddressShift += 2;  // one int (2 bytes) for address
    }
    // EEPROMWriteInt(SERIAL_BUFF, (unsigned int)i2cEepromAddress);  //save the I2C EEPROM tail address for saving a 'K' skill received from the serial port.
    //However, it will cause segfault when 'T' is called for the first time, because it has never been set by a 'K'. The following solution is used.
    EEPROMWriteInt(SERIAL_BUFF, theLastEEaddress);  //save the address of the last skill (named "zz") that is only used when calling token 'T' for the first time. the future 'K' command will overwrite it.
    PTL();
#ifdef I2C_EEPROM
    {
      PTF("Instincts used ");
      PT(INITIAL_SKILL_DATA_ADDRESS);
      PT('~');
      PT(i2cEepromAddress);
      PT('/');
      PT(EEPROM_SIZE);
      PTLF(" on I2C EEPROM\n");
    }
#endif
  }
#endif

  void assignSkillAddressToOnboardEeprom() {
    int skillAddressShift = 0;
    PTLF("Skill");
    for (byte s = 0; s < sizeof(progmemPointer) / 2; s++) {  //save skill info to on-board EEPROM, load skills to SkillList
      //        PTL(s);
      byte nameLen = EEPROM.read(SKILLS + skillAddressShift++);  //without last type character
      //    for (int n = 0; n < nameLen; n++)
      //      PT((char)EEPROM.read(SKILLS + skillAddressShift + n));
      skillAddressShift += nameLen;
      char skillType = EEPROM.read(SKILLS + skillAddressShift++);
      //    PT(skillType);
      //    PT('\t');
      //    PTL((int8_t)EEPROM.read(SKILLS + skillAddressShift));
#ifdef I2C_EEPROM
      if (skillType == 'N')  // the address of I(nstinct) has been written in previous operation: saveSkillNameFromProgmemToOnboardEEPROM() in instinct.ino
                             // if skillType == N(ewbility), save pointer address of progmem data array to onboard eeprom.
                             // it has to be done for different sketches because the addresses are dynamically assigned
#endif
        EEPROMWriteInt(SKILLS + skillAddressShift + 1, (int)progmemPointer[s]);
#if defined(I2C_EEPROM) && defined(MAIN_SKETCH)
      else
        s--;
#endif
      skillAddressShift += 3;
    }
    //  PTLF("Load finished!");
  }

  int lookupAddressByName(const char* key) {
    int skillAddressShift = 0;
    byte nSkills = eeprom(NUM_SKILLS);
    byte randSkillIdx = strcmp(key, "x") ? nSkills : random(nSkills);
    byte keyLen = strlen(key);
    char lr = key[keyLen - 1];
    for (byte s = 0; s < nSkills; s++) {  //save skill info to on-board EEPROM, load skills to SkillList
      byte nameLen = EEPROM.read(SKILLS + skillAddressShift++);
      // char* readName = new char[nameLen + 1];
      char readName[CMD_LEN + 1];
      for (byte l = 0; l < nameLen; l++) {
        readName[l] = EEPROM.read(SKILLS + skillAddressShift++);
      }
      readName[nameLen] = '\0';
#ifndef GYRO_PIN
      if (s == randSkillIdx && (!strcmp(readName, "bf") || !strcmp(readName, "ff") || !strcmp(readName, "rc") || !strcmp(readName, "rl") || !strcmp(readName, "jmp"))) {  //forbid violent motions in random mode
        randSkillIdx++;
        continue;
      }
#endif
      if (s == randSkillIdx          //random skill
          || !strcmp(readName, key)  //exact match: gait type + F or L, behavior
          // || readName[nameLen - 1] == 'L' && !strncmp(readName, key, nameLen - 1)  //gait type + R
          || readName[nameLen - 1] != 'F' && strcmp(readName, "bk") && !strncmp(readName, key, keyLen - 1) && (lr == 'L' || lr == 'R' || lr == 'X')  // L, R or X
      ) {
        // delete[] readName;
        period = EEPROM.read(SKILLS + skillAddressShift + 1);
        int bufferLen = dataLen(period);
        return SKILLS + skillAddressShift;
      }
      // delete[] readName;
      skillAddressShift += 4;  //1 byte type, 1 byte period, 1 int address
    }
    PT('?');  //key not found
    PTL(key);
    return -1;
  }

  int dataLen(int8_t p) {
    skillHeader = p > 0 ? 4 : 7;
    frameSize = p > 1 ? WALKING_DOF :  //gait
                  p == 1 ? DOF
                         :  //posture
                  DOF + 4;  //behavior
    int len = skillHeader + abs(p) * frameSize;
    return len;
  }

  void loadDataFromProgmem(unsigned int pgmAddress) {
    //      newCmd = new int8_t [bufferLen + 1];
    //      newCmd[0] = pgm_read_byte(pgmAddress++);
    int bufferLen = dataLen(period);
    for (int i = 0; i < bufferLen; i++)
      newCmd[i] = pgm_read_byte(pgmAddress++);
    //      newCmd[bufferLen] = '\0';
  }

  void loadDataFromI2cEeprom(unsigned int eeAddress) {
    Wire.beginTransmission(DEVICE_ADDRESS);
    Wire.write((int)((eeAddress) >> 8));  // MSB
    Wire.write((int)((eeAddress)&0xFF));  // LSB
    Wire.endTransmission();
    Wire.requestFrom((uint8_t)DEVICE_ADDRESS, (uint8_t)1);
    period = newCmd[0] = Wire.read();
    int bufferLen = dataLen(period);
    //      int tail = bufferLen;
    int readFromEE = 0;
    int readToWire = 0;
    while (bufferLen > 0) {
      //PTL("request " + String(min(WIRE_BUFFER, len)));
      Wire.requestFrom((uint8_t)DEVICE_ADDRESS, (uint8_t)min(WIRE_BUFFER, bufferLen));
      readToWire = 0;
      do {
        if (Wire.available()) newCmd[1 + readFromEE++] = Wire.read();
        /*PT( (int8_t)dutyAngles[readFromEE - 1]);
            PT('\t')*/
      } while (--bufferLen > 0 && ++readToWire < WIRE_BUFFER);
      //PTL();
    }
    //      newCmd[tail] = '\0';
  }
  void inplaceShift() {
    int angleLen = abs(period) * frameSize;                 // need one extra byte for terminator '~'
    int shiftRequiredByNewCmd = CMD_LEN - skillHeader + 1;  // required shift to store CMD_LEN + 1 chars. it can hold a command with CMD_LEN chars. the additioanl byte is required by '\0'.
    spaceAfterStoringData = BUFF_LEN - angleLen - 1;        // the bytes before the dutyAngles. The allowed command's bytes needs to -1
    // PTH("request", shiftRequiredByNewCmd);
    // PTH("aloShft", BUFF_LEN - (skillHeader + angleLen));
    // if (CMD_LEN > spaceAfterStoringData)
    //   PTF("LMT");
    // PTL(spaceAfterStoringData);
    for (int i = 0; i <= angleLen; i++)
      newCmd[BUFF_LEN - i] = newCmd[skillHeader + angleLen - i];
    newCmd[BUFF_LEN] = '~';
    dutyAngles = (int8_t*)newCmd + BUFF_LEN - angleLen;
  }
  void formatSkill() {
    transformSpeed = 2;
    for (int i = 0; i < 2; i++) {
      expectedRollPitch[i] = (int8_t)newCmd[1 + i];
      yprTilt[2 - i] = 0;
    }
    angleDataRatio = newCmd[3];
    byte baseHeader = 4;
    if (period < 0) {
      for (byte i = 0; i < 3; i++)
        loopCycle[i] = newCmd[baseHeader + i];
    }
#ifdef POSTURE_WALKING_FACTOR
    postureOrWalkingFactor = (period == 1 ? 1 : POSTURE_WALKING_FACTOR);
#endif
    firstMotionJoint = (period <= 1) ? 0 : DOF - WALKING_DOF;
    // dutyAngles = newCmd + skillHeader;
    inplaceShift();
#ifdef DEVELOPER
    info();
#endif
  }

  //    void skillByName(const char* skillName, byte angleDataRatio = 1, float speedRatio = 1, bool shutServoAfterward = true) {
  //      loadFrame(skillName);
  //      transform(dutyAngles, angleDataRatio, speedRatio);
  //      if (shutServoAfterward) {
  //        pwm.shutServos();
  //        token = 'd';
  //      }
  //    }

  void loadFrameByDataBuffer() {
    //      period = newCmd[0];
    //      dataLen(period);
    formatSkill();
    frame = 0;
    transform(dutyAngles + frame * frameSize, angleDataRatio, transformSpeed, firstMotionJoint);
  }

  void loadFrame(const char* skillName) {  //get lookup information from on-board EEPROM and read the data array from storage
    char lr = skillName[strlen(skillName) - 1];
    offsetLR = (lr == 'L' ? 25 : (lr == 'R' ? -25 : 0));
    int onBoardEepromAddress = lookupAddressByName(skillName);
    if (onBoardEepromAddress != -1) {
      char skillType = EEPROM.read(onBoardEepromAddress);  //load data by onboard EEPROM address
      unsigned int dataArrayAddress = EEPROMReadInt(onBoardEepromAddress + 2);
      char tempName[CMD_LEN + 1];
      strcpy(tempName, skillName);
#ifdef I2C_EEPROM
      if (skillType == 'I')  //copy instinct data array from external i2c eeprom
        loadDataFromI2cEeprom(dataArrayAddress);
      else  //copy newbility data array from progmem
#endif
        loadDataFromProgmem(dataArrayAddress);
      formatSkill();
      strcpy(newCmd, tempName);
      if (strcmp(newCmd, "calib") && period == 1)
        protectiveShift = random() % 100 / 10.0 - 5;
      else
        protectiveShift = 0;
      for (byte i = 0; i < DOF; i++)
        dutyAngles[i] += protectiveShift;
      // info();
      if (lr == 'R' || (lr == 'X' || lr != 'L') && random(100) % 2)
        mirror();
      frame = 0;
      transform(dutyAngles + frame * frameSize, angleDataRatio, transformSpeed, firstMotionJoint);
#ifdef NYBBLE
      for (byte i = 0; i < HEAD_GROUP_LEN; i++)
        targetHead[i] = currentAng[i];
#endif
    }
  }
  void mirror() {
    expectedRollPitch[0] = -expectedRollPitch[0];
    for (int k = 0; k < abs(period); k++) {
      if (period <= 1) {
        dutyAngles[k * frameSize] = -dutyAngles[k * frameSize];
        dutyAngles[k * frameSize + 2] = -dutyAngles[k * frameSize + 2];
      }
      for (byte col = (period > 1) ? 0 : 2; col < ((period > 1) ? WALKING_DOF : DOF) / 2; col++) {
        int8_t temp = dutyAngles[k * frameSize + 2 * col];
        dutyAngles[k * frameSize + 2 * col] = dutyAngles[k * frameSize + 2 * col + 1];
        dutyAngles[k * frameSize + 2 * col + 1] = temp;
      }
    }
  }
#define PRINT_SKILL_DATA
  void info() {
    PTF("prd:");
    PT(period);
    PT(",\texpct(pitch,roll):");
    PT(expectedRollPitch[0]);
    PT(",");
    PT(expectedRollPitch[1]);
    PT("\t");
    PTF("ratio: ");
    PTL(angleDataRatio);
    if (period < 0) {
      PT("loop: ");
      for (byte i = 0; i < 3; i++)
        PT(String((byte)loopCycle[i]) + ", ");
      PTL();
    }
#ifdef PRINT_SKILL_DATA
    for (int k = 0; k < abs(period); k++) {
      for (int col = 0; col < frameSize; col++) {
        PT((int8_t)dutyAngles[k * frameSize + col]);
        PT("\t");
      }
      PTL();
    }
#endif
    PTL();
  }

  void convertTargetToPosture(int* targetFrame) {
    int extreme[2];
    getExtreme(targetFrame, extreme);
    if (extreme[0] < -125 || extreme[1] > 125) {
      angleDataRatio = 2;
      for (int i = 0; i < DOF; i++)
        targetFrame[i] /= 2;
    } else
      angleDataRatio = 1;
    arrayNCPY(dutyAngles, targetFrame, DOF);
    period = 1;
    firstMotionJoint = 0;
    frameSize = DOF;
    frame = 0;
  }
  void perform() {
    if (period < 0) {  //behaviors
      int8_t repeat = loopCycle[2] >= 0 && loopCycle[2] < 2 ? 0 : loopCycle[2] - 1;
      for (byte c = 0; c < abs(period); c++) {  //the last two in the row are transition speed and delay
                                                //          PT("step "); PTL(c);
                                                //          printList(dutyAngles + c * frameSize);

        transform(dutyAngles + c * frameSize, angleDataRatio, dutyAngles[DOF + c * frameSize] / 4.0);

#ifdef GYRO_PIN  //if opt out the gyro, the calculation can be really fast
        if (dutyAngles[DOF + 2 + c * frameSize]) {
          int triggerAxis = dutyAngles[DOF + 2 + c * frameSize];
          int triggerAngle = dutyAngles[DOF + 3 + c * frameSize];

          float currentYpr = ypr[abs(triggerAxis)];
          float previousYpr = currentYpr;
          //            long triggerTimer = millis();
          while (1) {
            read_IMU();
            currentYpr = ypr[abs(triggerAxis)];
            //              print6Axis();
            //              PT(currentYpr);
            //              PT('\t');
            //              PTL(triggerAngle);
            if ((180 - fabs(currentYpr) > 2)                                                                                           //skip the angle when the reading jumps from 180 to -180
                && (triggerAxis * currentYpr > triggerAxis * triggerAngle && triggerAxis * previousYpr < triggerAxis * triggerAngle))  //the sign of triggerAxis will deterine whether the current angle should be larger or smaller than the trigger angle
              break;
            previousYpr = currentYpr;
          }
        }
#endif
        delay(abs(dutyAngles[DOF + 1 + c * frameSize] * 50));
        if (repeat != 0 && c != 0 && c == loopCycle[1]) {
          c = loopCycle[0] - 1;
          if (repeat > 0)  //if repeat <0, infinite loop. only the reset button will break the loop
            repeat--;
        }
      }
    } else {          //postures and gaits
#if defined GYRO_PIN  //&& !defined RANDOM_MIND
      // if (imuUpdated)
      if (gyroBalanceQ && !(frame % imuSkip)) {
        for (byte i = 0; i < 2; i++) {
          RollPitchDeviation[i] = ypr[2 - i] - expectedRollPitch[i];                                                                          //all in degrees
          RollPitchDeviation[i] = sign(ypr[2 - i]) * max(float(fabs(RollPitchDeviation[i]) - levelTolerance[i]), float(0)) + yprTilt[2 - i];  //filter out small angles
        }
      }
#endif
      if (jointIndex >= DOF) {
        jointIndex = 0;
        // timer = (timer + 1) % abs(skill.period);
        frame += tStep;
        if (frame >= abs(period)) {
          frame = 0;
        }
      }
#ifndef HEAD
      if (jointIndex == 0)
        jointIndex = 2;
#endif
#ifndef TAIL
      if (jointIndex == 2)
        jointIndex = DOF - WALKING_DOF;
#endif
#if WALKING_DOF == 8
      if (jointIndex == 4)
        jointIndex = 8;
#endif
      float duty;
      if (abs(period) > 1 && jointIndex < firstMotionJoint || abs(period) == 1 && jointIndex < 4 && manualHeadQ) {
        if (!manualHeadQ && jointIndex < 4) {
          duty = (jointIndex != 1 ? offsetLR : 0)  //look left or right
                 + 10 * sin(frame * (jointIndex + 2) * M_PI / abs(period));
        } else {
          duty =
#ifdef DAMPED_MOTION
            currentAng[jointIndex] + max(-10, min(10, (targetHead[jointIndex] - currentAng[jointIndex])))
#else
            targetHead[jointIndex]
#endif
            - gyroBalanceQ * currentAdjust[jointIndex];
        }
      } else
        duty = dutyAngles[frame * frameSize + jointIndex - firstMotionJoint] * angleDataRatio;
      calibratedPWM(jointIndex, duty
#if defined GYRO_PIN  //&& !defined RANDOM_MIND
                                  + gyroBalanceQ * (!exceptions ? (!(frame % imuSkip) ? adjust(jointIndex) : currentAdjust[jointIndex]) : 0)
#endif
      );
      jointIndex++;
    }
  }
};

Skill skill;

#ifdef BITTLE
int8_t expect[] = {
  57,
  43,
  60,
  47,
  -18,
  7,
  -17,
  7,
};
#elif defined NYBBLE
int8_t expect[] = {
  51,
  39,
  -57,
  -43,
  -18,
  7,
  19,
  -7,
};
#endif

int testEEPROM(char* skillData) {
  skill.loadFrame(skillData);
  PTL(skill.period);
  int len = 0;
  while (len < 8) {
    if (skill.dutyAngles[len] != expect[len])
      return 0;
    PT(int8_t(skill.dutyAngles[len]));
    PT('\t');
    len++;
  }
  return 1;
}

#ifndef MAIN_SKETCH
void writeConst() {
  //  flushEEPROM();
  int melodyAddress = MELODY_NORMAL;
  saveMelody(melodyAddress, melodyNormalBoot, sizeof(melodyNormalBoot));
  saveMelody(melodyAddress, melodyInit, sizeof(melodyInit));
  saveMelody(melodyAddress, melodyLowBattery, sizeof(melodyLowBattery));
  // saveMelody(melodyAddress, melody1, sizeof(melody1));
  EEPROM.update(BOOTUP_SOUND_STATE, 1);  //always turn on the boot up sound after reset
#if !defined AUTO_INIT
  playMelody(MELODY_INIT);
#endif
#ifndef AUTO_INIT
  PTLF("Reset joint offsets?(Y/n)");  //(Input ‘Y’ and hit enter, if you want to reset all the joint offsets to 0)
  char resetJointCalibrationQ = getUserInputChar();
  PTL(resetJointCalibrationQ);
#endif
  for (byte i = 0; i < DOF; i++) {
#ifndef AUTO_INIT
    if (resetJointCalibrationQ == 'Y' || resetJointCalibrationQ == 'y')
#endif
      EEPROM.update(CALIB + i, servoCalib[i]);
    EEPROM.update(PWM_PIN + i, pwm_pin[i]);
    EEPROM.update(MID_SHIFT + i, middleShift[i]);
    EEPROM.update(ROTATION_DIRECTION + i, rotationDirection[i]
#ifdef INVERSE_SERVO_DIRECTION
                                            * -1
#endif
    );
    for (byte j = 0; j < 2; j++) {
      EEPROM.update(ADAPT_PARAM + i * 2 + j, (int8_t)round(adaptiveParameterArray[i][j]));
      EEPROMWriteInt(ANGLE_LIMIT + i * 4 + j * 2, angleLimit[i][j]);
    }
    delay(6);
  }
  //PTL();
}

int configureEEPROM() {
  PTLF("\n* Change model and board definitions in OpenCat.ino!");
  PTLF("\nConfigure EEPROM");
  writeConst();  // only run for the first time when writing to the board.
  wdt_enable(WDTO_8S);
  skill.saveSkillInfoFromProgmemToOnboardEeprom();
  wdt_reset();
  return 1;
}
#endif
